Description: Add version info for Ubuntu.
 Note(coreycb): The accompanying keystone.conf update to
 [extra_headers] that sets "Distribution = Ubuntu" has
 been from this patch to debian/keystone.conf.dist.
Author: Yolanda Robla <yolanda.robla@canonical.com>
--- a/keystone/conf/__init__.py
+++ b/keystone/conf/__init__.py
@@ -30,6 +30,7 @@
 from keystone.conf import endpoint_filter
 from keystone.conf import endpoint_policy
 from keystone.conf import eventlet_server
+from keystone.conf import extra_headers
 from keystone.conf import federation
 from keystone.conf import fernet_receipts
 from keystone.conf import fernet_tokens
@@ -68,6 +69,7 @@
     endpoint_filter,
     endpoint_policy,
     eventlet_server,
+    extra_headers,
     federation,
     fernet_receipts,
     fernet_tokens,
--- /dev/null
+++ b/keystone/conf/extra_headers.py
@@ -0,0 +1,36 @@
+# Licensed under the Apache License, Version 2.0 (the "License"); you may
+# not use this file except in compliance with the License. You may obtain
+# a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+# WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+# License for the specific language governing permissions and limitations
+# under the License.
+
+from oslo_config import cfg
+
+from keystone.conf import utils
+
+
+distribution = cfg.StrOpt(
+    'Distribution',
+    default='Ubuntu',
+    help=utils.fmt("""
+Specifies the distribution of the keystone server.
+"""))
+
+GROUP_NAME = __name__.split('.')[-1]
+ALL_OPTS = [
+    distribution,
+]
+
+
+def register_opts(conf):
+    conf.register_opts(ALL_OPTS, group=GROUP_NAME)
+
+
+def list_opts():
+    return {GROUP_NAME: ALL_OPTS}
--- a/keystone/server/flask/request_processing/middleware/auth_context.py
+++ b/keystone/server/flask/request_processing/middleware/auth_context.py
@@ -127,6 +127,11 @@
         headers = list(headers)
     headers.append(('Vary', 'X-Auth-Token'))
 
+    # extra headers
+    for key, value in CONF.extra_headers.items():
+        if value != 'None':
+            headers.append(('X-'+key, value))
+
     if body is None:
         body = b''
         status = status or (http.client.NO_CONTENT,
